torch/csrc/jit/python/python_sugared_value.cpp

```diff
@@ -1202,6 +1222,12 @@ std::shared_ptr<SugaredValue> toSugaredValue(
     if (obj_name.compare("torch.cuda") == 0) {
       return std::make_shared<CUDAPythonModuleValue>(obj);
     }
+    // we register two APIs device_count and current_device in
+    // torch.backends.xpu for two reasons: 1) they are used to support JIT
+    // script. 2) it is independent of xpu extension.
```

Comment:
I don't think I follow this. Could you explain what these new API's are for?

If `device_count()` and `current_device()` are meant to be generic ops, then they should be registered in a way that's unrelated to XPU.

If they're XPU specific ops though, then you should be able to register them out of core, as part of the extension!

Reply:
Actually, I have to register xpu.device_count() and xpu.current_device() in JIT/TorchScript. Because if I don't register them, get_current_device_index() in _C/_utils.py will raise an error when get_current_device_index is in JIT and PyTorch is only running in a CUDA backend.
Similarly, I have to register cuda.device_count() and cuda.current_device() in PyTorch. Because It makes get_current_device_index available in JIT and running in an XPU backend.

Reply:
@davidberard98, we would like to add xpu backend support in _get_device_index. Before this PR, calls to cuda.device_count() and get_current_device_index() cannot be scripted in XPU backend if PyTorch is compiled without CUDA.

Reply:
@guangyey another question, do you really need get_current_device_index to be scriptable? As your comments point out, it seems like it's not generally scriptable except when cuda is currently built so it's not completely scriptable as it is now.

okay, my understanding is:
* if you want to be able to convert torch.backends.xpu.device_count() to an op (`xpu::device_count`), then you need to add this code in python_sugared_value.
* To get these ops out of core instead, you'd have to make a new torch library with ops that have these functions, and call those ops directly. (at least that's the way I imagine it would work)
* ^ That isn't really a viable option if you want to be able to script get_current_device_index and use it in the xpu context
* in addition, if we want `get_current_device_index` to be able to return the current device index for both CUDA and XPU backends, we need cuda::device_index and xpu::device_index to both be scriptable in either situation.

anyway this PR seems somewhat reasonable to me, but I defer the decision back to @bdhirsh since there's probably context about XPU that I'm missing.

Reply:
@bdhirsh @davidberard98 Thanks for your context and reviews.
Actually, we just want to enable native Python API _get_device_index() to be available on the XPU backend in both Eager mode and Script mode. This API is widely used in XPU extension and other third-party libs. And we need it can be compatible with CUDA, XPU, or other backends. So we have to support get_current_device_index () to be scripted() on the XPU backend. In addition, we still should guarantee that cuda::device_index and xpu::device_index are scriptable in each situation(CUDA, XPU, or other backends). This is the reason why we register cuda.device_count() and xpu.device_count() in PyTorch rather than out of the core.